File: Helpers/HostSpec.cs
namespace Wormhole.Helpers;

public readonly record struct HostSpec(string? User, string Host, int? Port);

public static class HostSpecParser
{
    /// <summary>
    /// Parses a connection target in any of the forms:
    /// <code>host</code>, <code>host:port</code>, <code>user@host</code>, <code>user@host:port</code>,
    /// <code>[ipv6]:port</code>, <code>user@[ipv6]:port</code>, or a bare bracketless IPv6 literal.
    /// </summary>
    public static HostSpec Parse(string input)
    {
        ArgumentNullException.ThrowIfNull(input);
        var s = input.AsSpan().Trim();

        string? user = null;
        var at = s.IndexOf('@');
        if (at > 0)
        {
            user = s[..at].ToString();
            s = s[(at + 1)..];
        }

        // Bracketed form: [host]:port — host may contain colons (IPv6).
        if (!s.IsEmpty && s[0] == '[')
        {
            var closeBracket = s.IndexOf(']');
            if (closeBracket < 0)
            {
                throw new FormatException(
                    $"Invalid host specifier '{input}': missing ']' to close bracketed host.");
            }
            var host = s[1..closeBracket];
            var rest = s[(closeBracket + 1)..];
            if (host.IsEmpty)
            {
                throw new FormatException(
                    $"Invalid host specifier '{input}': bracketed host is empty.");
            }
            if (rest.IsEmpty)
            {
                return new HostSpec(user, host.ToString(), null);
            }
            if (rest.Length > 1 && rest[0] == ':' && TryParsePort(rest[1..], out var bp))
            {
                return new HostSpec(user, host.ToString(), bp);
            }
            // Reject garbage after the closing bracket so callers don't silently
            // connect to the wrong endpoint with the trailing chars stripped.
            throw new FormatException(
                $"Invalid host specifier '{input}': unexpected characters after ']'.");
        }

        // Bracketless IPv6 literal: more than one ':' means we can't tell which is the port
        // separator; treat the whole string as host and let the caller fall back to the
        // protocol default port. Users who want an explicit IPv6 port must bracket the host.
        if (CountChar(s, ':') > 1)
        {
            if (s.IsEmpty)
                throw new FormatException($"Invalid host specifier '{input}': host is empty.");
            return new HostSpec(user, s.ToString(), null);
        }

        int? port = null;
        var colon = s.LastIndexOf(':');
        if (colon > 0 && TryParsePort(s[(colon + 1)..], out var p))
        {
            port = p;
            s = s[..colon];
        }

        if (s.IsEmpty)
        {
            throw new FormatException(
                $"Invalid host specifier '{input}': host is empty.");
        }

        return new HostSpec(user, s.ToString(), port);
    }

    private static bool TryParsePort(ReadOnlySpan<char> text, out int port)
    {
        if (int.TryParse(text, out port) && port >= 1 && port <= 65535)
        {
            return true;
        }
        port = 0;
        return false;
    }

    private static int CountChar(ReadOnlySpan<char> s, char c)
    {
        var n = 0;
        for (var i = 0; i < s.Length; i++)
        {
            if (s[i] == c) n++;
        }
        return n;
    }
}

File: rust/crates/wormhole-ui/src/connection_editor/host_spec.rs
//! Minimal `HostSpecParser` port for HTTP address splitting.

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HostSpec {
    pub host: String,
    pub port: Option<i32>,
}

/// Parse `host`, `host:port`, `[ipv6]:port`, or a bare IPv6 literal.
pub fn parse(input: &str) -> Result<HostSpec, ()> {
    let s = input.trim();
    if s.is_empty() {
        return Err(());
    }

    // Bracketed: [host]:port
    if s.starts_with('[') {
        let close = s.find(']').ok_or(())?;
        let host = &s[1..close];
        if host.is_empty() {
            return Err(());
        }
        let rest = &s[close + 1..];
        if rest.is_empty() {
            return Ok(HostSpec {
                host: host.to_string(),
                port: None,
            });
        }
        if let Some(port_text) = rest.strip_prefix(':')
            && let Some(port) = parse_port(port_text)
        {
            return Ok(HostSpec {
                host: host.to_string(),
                port: Some(port),
            });
        }
        return Err(());
    }

    // Bracketless IPv6: more than one ':' → whole string is host.
    if s.chars().filter(|c| *c == ':').count() > 1 {
        return Ok(HostSpec {
            host: s.to_string(),
            port: None,
        });
    }

    if let Some(colon) = s.rfind(':')
        && colon > 0
        && let Some(port) = parse_port(&s[colon + 1..])
    {
        let host = &s[..colon];
        if host.is_empty() {
            return Err(());
        }
        return Ok(HostSpec {
            host: host.to_string(),
            port: Some(port),
        });
    }

    Ok(HostSpec {
        host: s.to_string(),
        port: None,
    })
}

fn parse_port(text: &str) -> Option<i32> {
    let port: i32 = text.parse().ok()?;
    (1..=65535).contains(&port).then_some(port)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn host_port_and_ipv6() {
        assert_eq!(
            parse("10.0.0.1:8443").unwrap(),
            HostSpec {
                host: "10.0.0.1".into(),
                port: Some(8443),
            }
        );
        assert_eq!(
            parse("[fd00::1]:443").unwrap(),
            HostSpec {
                host: "fd00::1".into(),
                port: Some(443),
            }
        );
        assert_eq!(
            parse("fd00::1").unwrap(),
            HostSpec {
                host: "fd00::1".into(),
                port: None,
            }
        );
    }
}
